gitrevise/todo.py

```diff
@@ -242,27 +242,63 @@ def edit_todos(
     return result
 
 
+def is_fixup(todo: Step) -> bool:
+    return todo.kind in (StepKind.FIXUP, StepKind.SQUASH)
+
+
+def squash_message_template(
+    target_message: bytes, fixups: List[Tuple[StepKind, bytes]]
+) -> bytes:
+    fused = (
+        b"# This is a combination of %d commits.\n" % (len(fixups) + 1)
+        + b"# This is the 1st commit message:\n"
+        + b"\n"
+        + target_message
+    )
+
+    for index, (kind, message) in enumerate(fixups):
+        fused += b"\n"
+        if kind == StepKind.FIXUP:
+            fused += (
+                b"# The commit message #%d will be skipped:\n" % (index + 2)
+                + b"\n"
+                + b"".join(b"# " + line for line in message.splitlines(keepends=True))
+            )
+        else:
+            assert kind == StepKind.SQUASH
+            fused += b"# This is the commit message #%d:\n\n%s" % (index + 2, message)
+
+    return fused
+
+
 def apply_todos(
     current: Optional[Commit],
     todos: List[Step],
     reauthor: bool = False,
 ) -> Commit:
-    for step in todos:
+    fixups: List[Tuple[StepKind, bytes]] = []
+
+    for index, step in enumerate(todos):
         rebased = step.commit.rebase(current).update(message=step.message)
         if step.kind == StepKind.PICK:
             current = rebased
-        elif step.kind == StepKind.FIXUP:
+        elif is_fixup(step):
             if current is None:
                 raise ValueError("Cannot apply fixup as first commit")
+            if not fixups:
+                fixup_target_message = current.message
+            fixups.append((step.kind, rebased.message))
             current = current.update(tree=rebased.tree())
+            is_last_fixup = index + 1 == len(todos) or not is_fixup(todos[index + 1])
+            if is_last_fixup:
+                if any(kind == StepKind.SQUASH for kind, message in fixups):
+                    current = current.update(
+                        message=squash_message_template(fixup_target_message, fixups)
+                    )
+                    current = edit_commit_message(current)
+                fixups.clear()
```

Comment:
Suggestion: Since you need to look ahead of at least squash and fixup steps, to see if it was the last, you might check this unconditionally for all step kinds for the benefit of discovering the beginning of the sequence in the iteration it happens. Then, the first commit could go into the same list as the rest instead of having a separate variable. The same check negated:

    is_squashed_into = index + 1 < len(todos) and is_fixup(todos[index + 1])

Do you know what else could use this? If a later commit that is squashed into this commit has a known tree (#133), we can also avoid raising any conflict on this commit, because this commit's tree is to be discarded.

A use case for this is when part of commit A belongs in commit B, so you carve out a fixup of B from A, which now comes before B (to preserve the metadata of the right commit, you may not want to simply squash B into "fixup B").